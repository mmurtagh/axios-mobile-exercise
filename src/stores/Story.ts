import { makeAutoObservable, toJS } from 'mobx';
import dayjs from 'dayjs';

export enum SupportedEntityType {
  LINK = 'LINK',
}

export enum SupportedBlockType {
  UNSTYLED = 'unstyled',
  UNORDERED_LIST_ITEM = 'unordered-list-item',
}

export enum SupportedInlineStyle {
  BOLD = 'BOLD',
}

/** 
 * @name sanitizeContentInstance
 * @description: Filters out unsupported features from raw AxiosContentInstances.
 * @param _contentInstance: The AxiosContentInstance object that needs to be sanitized.
*/
function sanitizeContentInstance (_contentInstance: AxiosContentInstance): AxiosContentInstance {
  const contentInstance = { ..._contentInstance }
  const { blocks: { blocks } } = contentInstance;

  // Filter out all blocks with unsupported block type
  contentInstance.blocks.blocks = blocks.filter(({ type }) => {
    const supportedTypes: string[] = Object.values(SupportedBlockType);

    return supportedTypes.includes(type)
  })

  return contentInstance
}

/** 
 * @class: Story
 * @description: An Axios story. A mobx domain object managed by StoryStore.js.
*/
export class Story {
  contentInstance: AxiosContentInstance;

  constructor(contentInstance: AxiosContentInstance) {
    this.contentInstance = sanitizeContentInstance(contentInstance);
    makeAutoObservable(this);
  }

  get headline(): string {
    return this.contentInstance.headline;
  }

  get publishedDate(): dayjs.Dayjs {
    return dayjs(this.contentInstance.published_date);
  }

  /** 
   * @name getImageSources
   * @description: Gets the list of image sources of a given ratio.
   * @param ratio: The desired image's aspect ratio
  */
   getImageSources (ratio: '1x1' | '4x3' | '16x9'): Crop[] {
    let image = this.contentInstance.primary_image

    if (image === null || !image.crops[ratio].sizes) {
      image = this.contentInstance.social_image;
    }

    if (image === null) {
      return []
    }

    return toJS(image.crops[ratio].sizes)
  }

  get id(): string {
    return this.contentInstance.id;
  }

  get author(): string | null {
    if (!this.contentInstance.authors.length) {
      return null;
    }

    return this.contentInstance.authors[0].display_name;
  }

  get primaryTopicName(): string | null {
    const topics = this.contentInstance.topics;

    if (topics === null || topics.length === 0) {
      return null;
    }

    return topics[0].name;
  }

  get blocks(): Block[] {
    return this.contentInstance.blocks.blocks;
  }

  get entityMap(): Entity[] {
    return this.contentInstance.blocks.entityMap;
  }

  /** 
   * @name imageDescription
   * @description: Gets a description of the image associated with the story.
  */
  get imageDescription(): string {
    let image = this.contentInstance.primary_image

    if (image === null) {
      image = this.contentInstance.social_image;
    }

    if (image === null) {
      return ''
    }

    const { blocks } = image.caption
    
    if (blocks.length === 0) {
      return ''
    }

    return blocks[0].text;
  }
}
