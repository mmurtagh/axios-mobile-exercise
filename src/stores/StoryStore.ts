import { makeAutoObservable, runInAction } from 'mobx';

import { AxiosService } from '../services/AxiosService';
import { Story } from './Story';

/** 
 * @class: StoryStore
 * @description: A mobx domain store representing a store of Axios stories.
*/
export class StoryStore {
  service: AxiosService;
  isLoadingStories = false;
  stories: Story[] = [];

  constructor() {
    makeAutoObservable(this);
    this.service = new AxiosService();
  }

  /** 
   * @class: loadMostRecentStories
   * @description: Loads the most recent stories using the Axios stream and content APIs.
  */
  async loadMostRecentStories(): Promise<void> {
    const stream: AxiosStream = await this.service.getStream();

    const contentCalls = stream.results.map(id => {
      return this.service.getContent(id);
    });

    const data: AxiosContentInstance[] = await Promise.all(contentCalls);

    runInAction(() => {
      this.stories = data.map((story) => new Story(story))
    });
  }

  getStory(id: string): Story | null {
    return (
      this.stories.find(story => {
        return id === story.id;
      }) || null
    );
  }
}
