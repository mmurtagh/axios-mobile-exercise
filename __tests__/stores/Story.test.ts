import dayjs from "dayjs";
import { Story } from "../../src/stores/Story";
import contentInstance from '../data/contentInstance';

describe('Story', () => {
  test('filters out unsupported blocks', () => {
    const story = new Story(contentInstance);
    const blocks = story.contentInstance.blocks.blocks;

    expect(blocks.length).toBe(2);
    expect(blocks.find(({ key }) => key === 'GAB')).toBeFalsy();
  });

  test('headline', () => {
    const story = new Story(contentInstance);

    expect(story.headline).toBe(contentInstance.headline);
  });

  test('publishedDate', () => {
    const story = new Story(contentInstance);
    const publishedDate = dayjs(contentInstance.published_date);

    expect(story.publishedDate.isSame(publishedDate)).toBeTruthy();
  });

  test('id', () => {
    const story = new Story(contentInstance);

    expect(story.id).toBe(contentInstance.id);
  });

  test('blocks', () => {
    const story = new Story(contentInstance);

    expect(story.blocks).toStrictEqual(contentInstance.blocks.blocks);
  });

  test('entityMap', () => {
    const story = new Story(contentInstance);

    expect(story.entityMap).toStrictEqual(contentInstance.blocks.entityMap);
  });

  describe('getImageSources', () => {
    test('gets image sources of correct ratio', () => {
      const story = new Story(contentInstance);

      expect(story.getImageSources('1x1')).toMatchSnapshot();
    });

    test('falls back to social_image if primary_image is null', () => {
      const value = { ...contentInstance };
      value.primary_image = null;
      value.social_image = contentInstance.primary_image;
      const story = new Story(contentInstance);

      expect(story.getImageSources('4x3')).toMatchSnapshot();
    })

    test('returns [] if neither primary_image or social_image exists', () => {
      const value = { ...contentInstance };
      value.primary_image = null;
      const story = new Story(value)

      const sources = story.getImageSources('1x1')

      expect(sources).toStrictEqual([]);
    });
  });

  describe('primaryTopicName', () => {
    test('returns first topic name', () => {
      const story = new Story(contentInstance);

      expect(story.primaryTopicName).toBe('Science');
    })

    test('returns null if topics = []', () => {
      const value = { ...contentInstance }
      value.topics = [];
      const story = new Story(value);

      expect(story.primaryTopicName).toBe(null);
    })

    test('returns null if topics = null', () => {
      const value = { ...contentInstance }
      value.topics = null;
      const story = new Story(value);

      expect(story.primaryTopicName).toBe(null);
    })
  })
})