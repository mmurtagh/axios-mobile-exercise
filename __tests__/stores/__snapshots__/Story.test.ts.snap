// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Story getImageSources falls back to social_image if primary_image is null 1`] = `
Array [
  Object {
    "height": 100,
    "url": "Image100-4x3",
    "width": 100,
  },
  Object {
    "height": 200,
    "url": "Image200-4x3",
    "width": 200,
  },
  Object {
    "height": 300,
    "url": "Image300-4x3",
    "width": 300,
  },
  Object {
    "height": 400,
    "url": "Image400-4x3",
    "width": 400,
  },
  Object {
    "height": 500,
    "url": "Image500-4x3",
    "width": 500,
  },
]
`;

exports[`Story getImageSources gets image sources of correct ratio 1`] = `
Array [
  Object {
    "height": 100,
    "url": "Image100-1x1",
    "width": 100,
  },
  Object {
    "height": 200,
    "url": "Image200-1x1",
    "width": 200,
  },
  Object {
    "height": 300,
    "url": "Image300-1x1",
    "width": 300,
  },
  Object {
    "height": 400,
    "url": "Image400-1x1",
    "width": 400,
  },
  Object {
    "height": 500,
    "url": "Image500-1x1",
    "width": 500,
  },
]
`;
